### PR TITLE
Add list subcommands for packages, channels, and bundles

### PR DESCRIPTION
**Pull request by joelanford**
Signed-off-by: Joe Lanford <joe.lanford@gmail.com>

**Description of the change:**
Add `opm alpha list packages|channels|bundles` subcommands

Examples:

`opm alpha list packages quay.io/openshift-community-operators/catalog:latest | head -10` 
```
NAME                                       DISPLAY NAME                                       DEFAULT CHANNEL
3scale-community-operator                  3scale API Management                              threescale-2.10
akka-cluster-operator                      Akka Cluster Operator                              alpha
alvearie-imaging-ingestion                 Alvearie Imaging Ingestion Operator                alpha
api-operator                               API Operator for Kubernetes                        2.x-stable
apicast-community-operator                 APIcast                                            stable
apicurio-registry                          Apicurio Registry Operator                         2.x
apicurito                                  Apicurito Operator                                 alpha
aqua                                       Aqua Security Operator                             6.2.0
argocd-operator                            Argo CD                                            alpha
```

`opm alpha list channels quay.io/openshift-community-operators/catalog:latest etcd`
```
PACKAGE  CHANNEL                HEAD
etcd     alpha                  etcdoperator.v0.6.1
etcd     clusterwide-alpha      etcdoperator.v0.9.4-clusterwide
etcd     singlenamespace-alpha  etcdoperator.v0.9.4
```

`opm alpha list bundles quay.io/openshift-community-operators/catalog:latest etcd`
```
PACKAGE  CHANNEL                BUNDLE                           REPLACES                         SKIPS  SKIP RANGE  IMAGE
etcd     alpha                  etcdoperator.v0.6.1                                                                  quay.io/openshift-community-operators/etcd:v0.6.1
etcd     clusterwide-alpha      etcdoperator.v0.9.0                                                                  quay.io/openshift-community-operators/etcd:v0.9.0
etcd     clusterwide-alpha      etcdoperator.v0.9.2-clusterwide  etcdoperator.v0.9.0                                 quay.io/openshift-community-operators/etcd:v0.9.2-clusterwide
etcd     clusterwide-alpha      etcdoperator.v0.9.4-clusterwide  etcdoperator.v0.9.2-clusterwide                     quay.io/openshift-community-operators/etcd:v0.9.4-clusterwide
etcd     singlenamespace-alpha  etcdoperator.v0.9.0                                                                  quay.io/openshift-community-operators/etcd:v0.9.0
etcd     singlenamespace-alpha  etcdoperator.v0.9.2              etcdoperator.v0.9.0                                 quay.io/openshift-community-operators/etcd:v0.9.2
etcd     singlenamespace-alpha  etcdoperator.v0.9.4              etcdoperator.v0.9.2                                 quay.io/openshift-community-operators/etcd:v0.9.4
```

**Motivation for the change:**
Assist users trying to understand the contents of a particular index.

NOTE: The `list` subcommands are meant for convenience and human consumption only. Users needing to build automation related to the contents of an index should use the `render` command to get the full JSON or YAML output of an index. The output format and fields of the `list` subcommands are _NOT_ an API and are subject to change without notice.

**Reviewer Checklist**
- [x] Implementation matches the proposed design, or proposal is updated to match implementation
- [x] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [x] Commit messages sensible and descriptive

### CODE REVIEW

File: internal/action/list_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+func TestListChannels(t *testing.T) {
+	type spec struct {
+		name     string
+		list     ListChannels
+		expected string
+	}
+	specs := []spec{
+		{
+			name: "Success/WithPackage",
```
Comment (awgreene) [inline internal/action/list_test.go:49]: Can we add a `Failure/WithPackage`?

File: internal/action/list_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+foo      stable   foo.v0.2.0  foo.v0.1.0  foo.v0.1.1,foo.v0.1.2  <0.2.0      test.registry/foo-operator/foo-bundle:v0.2.0
+`,
+		},
+		{
+			name: "Success/WithoutPackage",
+			list: ListBundles{IndexReference: "testdata/foo-index-v0.2.0-declcfg"},
+			expected: `PACKAGE  CHANNEL  BUNDLE      REPLACES    SKIPS                  SKIP RANGE  IMAGE
+foo      beta     foo.v0.1.0                                     <0.1.0      test.registry/foo-operator/foo-bundle:v0.1.0
+foo      beta     foo.v0.2.0  foo.v0.1.0  foo.v0.1.1,foo.v0.1.2  <0.2.0      test.registry/foo-operator/foo-bundle:v0.2.0
+foo      stable   foo.v0.2.0  foo.v0.1.0  foo.v0.1.1,foo.v0.1.2  <0.2.0      test.registry/foo-operator/foo-bundle:v0.2.0
```
Comment (awgreene) [inline internal/action/list_test.go:101]: It seems like this test might add more value if the expected result was not identical to the one immediately above. Could we add a non `foo` package to the foo index? If that is not the purpose of the foo index (IE, it should only contain foo packages), perhaps a new index?

**Review by awgreene — COMMENTED** [review]
Not exactly my area of expertise but it looks good to me. It would be nice if the command test cases capturing edge cases for the commands.

**Comment by joelanford** [comment]
Thanks for the review @awgreene. Updated the tests to cover more of the edge cases as well!

File: internal/declcfg/model_to_declcfg.go [hunk 3]
```diff
@@ -68,6 +68,14 @@ func traverseModelChannels(mpkg model.Package) []Bundle {
 	var out []Bundle
 	for _, b := range bundles {
 		b.Properties = property.Deduplicate(b.Properties)
+
+		sort.Slice(b.Properties, func(i, j int) bool {
+			if b.Properties[i].Type != b.Properties[j].Type {
+				return b.Properties[i].Type < b.Properties[j].Type
+			}
+			return string(b.Properties[i].Value) < string(b.Properties[j].Value)
+		})
```
Comment (joelanford) [inline internal/declcfg/model_to_declcfg.go:77]: I noticed some non-determinism in the tests because of the way that model bundles are traversed and combined when converting to declcfg bundles. To make conversions deterministic, we need to sort the properties.

Just highlighting this to explain that many of the property-related changes in the testdata are to reorder to match the expectations of the sort.

**Review by ankitathomas — APPROVED** [review]
/lgtm
